File: next-main/src/main/java/com/next/reservations/core/service/ReservationManagingService.java
package com.next.reservations.core.service;

import com.next.reservations.core.domain.Reservation;
import com.next.reservations.core.domain.ReservationDetails;
import com.next.reservations.core.domain.ReservationStatus;
import com.next.reservations.core.domain.ReservationTime;
import com.next.reservations.web.request.ReservationRequest;
import org.springframework.stereotype.Service;

import javax.transaction.Transactional;
import java.time.LocalDate;
import java.time.LocalTime;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

@Service
public class ReservationManagingService {

    private ReservationService reservationService;
    private ReservationDetailsService reservationDetailsService;
    private ReservationTimeService reservationTimeService;

    public ReservationManagingService(ReservationService reservationService,
                                      ReservationDetailsService reservationDetailsService,
                                      ReservationTimeService reservationTimeService) {
        this.reservationService = reservationService;
        this.reservationDetailsService = reservationDetailsService;
        this.reservationTimeService = reservationTimeService;
    }

    @Transactional
    public Reservation createReservation(ReservationRequest reservationRequest) {
        final ReservationDetails reservationDetails =
                reservationDetailsService.create(reservationRequest.fullName, reservationRequest.email,
                        reservationRequest.phoneNumber, reservationRequest.comment, reservationRequest.numberOfPlayers);

        final ReservationTime reservationTime = reservationTimeService.findById(reservationRequest.reservationTimeId);

        final List<Integer> dateParts = Arrays.stream(reservationRequest.date.split("/"))
                .mapToInt(Integer::parseInt).boxed().collect(Collectors.toList());

        final LocalDate localDate = LocalDate.of(dateParts.get(0), dateParts.get(1), dateParts.get(2));

        return reservationService.create(ReservationStatus.values()[reservationRequest.status],
                reservationDetails, reservationTime, localDate);
    }
}

File: next-main/src/main/kotlin/com/next/reservations/core/service/ReservationManagingService.kt
package com.next.reservations.core.service

import com.next.reservations.core.domain.Reservation
import com.next.reservations.core.domain.ReservationStatus
import com.next.reservations.core.utils.Utils
import com.next.reservations.web.request.ReservationRequest
import org.springframework.stereotype.Service
import java.time.LocalDate
import javax.transaction.Transactional

@Service
class ReservationManagingService(private val reservationService: ReservationService,
                                 private val reservationDetailsService: ReservationDetailsService,
                                 private val reservationTimeService: ReservationTimeService,
                                 private val reservationTimeConfigurationService: ReservationTimeConfigurationService) {

    @Transactional
    fun createReservation(reservationRequest: ReservationRequest): Reservation {

        val reservationDetails = reservationDetailsService.create(reservationRequest.fullName, reservationRequest.email,
                reservationRequest.phoneNumber, reservationRequest.comment, reservationRequest.numberOfPlayers)

        val reservationTime = reservationTimeService.findById(reservationRequest.reservationTimeId)

        val dateParts = reservationRequest.date.split("-").map { it.toInt() }

        val localDate = LocalDate.of(dateParts[0], dateParts[1], dateParts[2])

        return reservationService.create(ReservationStatus.values()[reservationRequest.status],
                reservationDetails, reservationTime, localDate)
    }


    fun findAllActiveReservationsByConfig(configId: Long): List<Reservation> {
        val reservationTimeIds = reservationTimeService.findAllByReservationTimeConfigId(configId).map { it.id }

        val reservations =
                reservationService.findAllByStatusInAndReservationTimeIdInAndReservationDateAfter(listOf(ReservationStatus.ACCEPTED, ReservationStatus.PENDING),
                        reservationTimeIds, LocalDate.now().minusDays(1))

        return reservations
    }

    fun findAllActiveReservationsByReservationTime(reservationTimeId: Long): List<Reservation> {
        val reservations =
                reservationService.findAllByStatusInAndReservationTimeIdInAndReservationDateAfter(listOf(ReservationStatus.ACCEPTED, ReservationStatus.PENDING),
                        listOf(reservationTimeId), LocalDate.now().minusDays(1))

        return reservations
    }

    fun findAllInvalidByNewFutureDefaultDate(date: LocalDate): List<Reservation> {

        val currentFutureDateConfig = reservationTimeConfigurationService.findFutureDefault()
        val defaultConfig = reservationTimeConfigurationService.findByDefaultConfigTrue()

        val futureDateReservations = if (currentFutureDateConfig != null) {
            val currentFutureDateReservationTimes = reservationTimeService.findAllByReservationTimeConfigId(currentFutureDateConfig.id).map { it.id }
            val futureDateReservations = reservationService.findAllByStatusInAndReservationTimeIdInAndReservationDateAfter(
                    listOf(ReservationStatus.ACCEPTED, ReservationStatus.PENDING), currentFutureDateReservationTimes, LocalDate.now())

            futureDateReservations
        } else listOf()

        val defaultReservationTimes = reservationTimeService.findAllByReservationTimeConfigId(defaultConfig.id).map{it.id}
        val defaultInvalidReservations = reservationService.findAllByStatusInAndReservationTimeIdInAndReservationDateAfter(
                listOf(ReservationStatus.ACCEPTED, ReservationStatus.PENDING), defaultReservationTimes, date.minusDays(1))

        return futureDateReservations.union(defaultInvalidReservations).toList()
    }

    @Transactional
    fun changeDefaultConfig(newDefaultConfigId: Long) {
        val currentDefault = reservationTimeConfigurationService.findByDefaultConfigTrue()
        val reservationsForRemoval = findAllActiveReservationsByConfig(currentDefault.id).map { it.id }
        reservationService.removeReservationsByIds(reservationsForRemoval)
        reservationTimeConfigurationService.setAsDefault(newDefaultConfigId)
    }

    @Transactional
    fun setNewFutureDefault(id: Long, newFutureDefaultDate: LocalDate) {
        val reservationsForRemoval = findAllInvalidByNewFutureDefaultDate(newFutureDefaultDate)
        reservationService.removeReservationsByIds(reservationsForRemoval.map { it.id })
        reservationTimeConfigurationService.removeFutureDefaultStartDate()
        reservationTimeConfigurationService.setNewFutureDefaultStartDate(id, newFutureDefaultDate)
    }

}
